src/data/list/rotate.lean

```diff
@@ -394,6 +434,160 @@ begin
+
+@[simp] lemma length_cyclic_permutations_cons (x : α) (l : list α) :
+  length (cyclic_permutations (x :: l)) = length l + 1 :=
+by simp [cyclic_permutations_cons]
+
+lemma length_cyclic_permutations_of_ne_nil (l : list α) (h : l ≠ []) :
+  length (cyclic_permutations l) = length l :=
+begin
+  obtain ⟨hd, tl, rfl⟩ := exists_cons_of_ne_nil h,
+  simp
+end
+
+@[simp] lemma nth_le_cyclic_permutations (l : list α) (n : ℕ)
+  (hn : n < length (cyclic_permutations l)) :
+  nth_le (cyclic_permutations l) n hn = l.rotate n :=
+begin
+  cases l with x l,
+  { simp },
+  { cases n,
+    { simp [init_eq_take, cyclic_permutations_cons] },
+    { suffices : drop n l ++ x :: take n l = (l ++ [x]).rotate n,
+      { simpa [init_eq_take, nth_le_take', cyclic_permutations_cons] },
+      replace hn : n < length l,
+      { simpa [succ_lt_succ_iff] using hn },
+      rw [rotate_eq_drop_append_take, drop_append_of_le_length hn.le,
+          take_append_of_le_length hn.le, append_assoc, singleton_append],
+      simpa using (hn.trans (lt_succ_self _)).le } }
+end
+
+lemma mem_cyclic_permutations_self (l : list α) :
+  l ∈ cyclic_permutations l :=
+begin
+  cases l with x l,
+  { simp },
+  { rw mem_iff_nth_le,
+    refine ⟨0, by simp, _⟩,
+    simp }
+end
+
+lemma length_mem_cyclic_permutations (l : list α) (h : l' ∈ cyclic_permutations l) :
+  length l' = length l :=
+begin
+  obtain ⟨k, hk, rfl⟩ := nth_le_of_mem h,
+  simp
+end
+
+lemma mem_cyclic_permutations_iff {l l' : list α} :
+  l ∈ cyclic_permutations l' ↔ l ~r l' :=
+begin
+  split,
+  { intro h,
+    obtain ⟨k, hk, rfl⟩ := nth_le_of_mem h,
+    simp },
+  { intro h,
+    obtain ⟨k, rfl⟩ := h.symm,
+    rw mem_iff_nth_le,
+    simp only [exists_prop, nth_le_cyclic_permutations],
+    cases l' with x l,
+    { simp },
+    { refine ⟨k % length (x :: l), _, rotate_mod _ _⟩,
+      simpa using nat.mod_lt _ (zero_lt_succ _) } }
+end
+
+@[simp] lemma cyclic_permutations_eq_nil_iff {l : list α} :
+  cyclic_permutations l = [[]] ↔ l = [] :=
+begin
+  refine ⟨λ h, _, λ h, by simp [h]⟩,
+  rw [←is_rotated_nil_iff', ←mem_cyclic_permutations_iff, h, mem_singleton]
+end
+
+@[simp] lemma cyclic_permutations_eq_singleton_iff {l : list α} {x : α} :
+  cyclic_permutations l = [[x]] ↔ l = [x] :=
+begin
+  refine ⟨λ h, _, λ h, by simp [cyclic_permutations, h, init_eq_take]⟩,
+  rw [←is_rotated_singleton_iff', ←mem_cyclic_permutations_iff, h, mem_singleton]
+end
+
+/-- If a `l : list α` is `nodup l`, then all of its cyclic permutants are distinct. -/
+lemma nodup.cyclic_permutations {l : list α} (hn : nodup l) :
+  nodup (cyclic_permutations l) :=
+begin
+  cases l with x l,
+  { simp },
+  rw nodup_iff_nth_le_inj,
+  intros i j hi hj h,
+  simp only [length_cyclic_permutations_cons] at hi hj,
+  rw [nth_le_cyclic_permutations, nth_le_cyclic_permutations, rotate_eq_iff,
+      rotate_rotate, eq_comm, hn.rotate_eq_self_iff, length_rotate, length,
+      mod_eq_of_lt hi] at h,
```

Comment:
I think there's a helper lemma missing here from the intermediate goal state:

```lean
hn: l.nodup
hj: j < l.length
hi: i < l.length
h: l.rotate i = l.rotate j
⊢ i = j
```
or possibly
```lean
hn: l.nodup
h: l.rotate i = l.rotate j
⊢ i % l.length = j % l.length
```